platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,670 @@
+it provides an overview of the whole system, with an emphasis on the Consensus module, and how the Consensus module
+interacts with the Execution module.
+
+This document supports existing features implemented in new ways, and it provides for new features (such as the dynamic
+address book) which have not been implemented. After acceptance, a long series of changes will be required to modify
+the existing codebase to meet this new design. This will not happen overnight, nor will it block progress on all other
+initiatives. Instead, this plan provides the blueprint for our new node architecture, which we will work towards
+implementing with every change we make going forward. This blueprint will also provide the framework within which we
+will evaluate all other feature designs and implementations.
+
+## Design
+
+![Network](network.png)
+
+A group of consensus nodes together form a _consensus network_. The network, as a whole, takes as input _transactions_
+and each node in the network produces as output a _blockchain_ represented as a _block stream_. Each node in the network
+maintains _state_. The job of a node is to (a) work with other nodes to come to _consensus_ on which transactions to
+include, (b) order those transactions and assign a _consensus timestamp_ to each transaction, (c) handle each
+transaction, transitioning state from _S_ to _S'_, (d) produce blocks containing a history of inputs and state
+transitions, (e) work with other nodes to sign each block, and (f) export the block.
+
+As a decentralized network, each node can make no assumptions about the other nodes in the network. Other nodes may be
+faster, or slower. They may have superior or inferior network connections. They may be far away from each other or
+colocated. Each of those parameters may change over time. They may be running modified software or the official builds.
+They may be honest, or dishonest. Each node must assume that just more than 2/3 of the other nodes are honest, but must
+also assume that any particular node may be dishonest.
+
+The network must also _as a whole_ remain resilient and operational regardless of the transaction load supplied to the
+network. Nodes that are unable to keep up with the transaction load must be able to fail gracefully and rejoin the
+network if conditions improve. If a sufficiently large number of nodes are unable to keep up with the transaction load,
+then the network as a whole must collaborate to _throttle_ transaction ingestion to a level that will permit the network
+to remain stable.
+
+The design of the consensus node _does not require_ careful tuning for particular execution environments in order
+to remain live and responsive. Indeed, it is a hard requirement that tuning _cannot be required_. It is infeasible to
+test the exact configuration of a decentralized network, by definition, and therefore cannot be required for stable
+operation.
+
+![Design](consensus-module-arch.png)
```

Comment:
In this diagram, I assume your intention is that gossip will receive the roster for a given round from the `Hashgraph` module. So, a roster update would look something like this:

execution -> hashgraph -> gossip

If this is the intention, I don't think it is correct. Gossip should receive rosters for future rounds ASAP, to be able to gossip as far into the future as possible. Gossip shouldn't wait to get the roster for a round until after that round has reached consensus.

If the execution layer is dictating which rosters to use, why bother routing the rosters through the `Hashgraph` module? Can't we just give a new roster to everyone that needs it in the consensus module, as soon as a roster is received from the execution module?

Reply:
> If this is the intention, I don't think it is correct. Gossip should receive rosters for future rounds ASAP, to be able to gossip as far into the future as possible. Gossip shouldn't wait to get the roster for a round until after that round has reached consensus.

Can you provide a good example of where this is meaningful?

> If the execution layer is dictating which rosters to use, why bother routing the rosters through the Hashgraph module? Can't we just give a new roster to everyone that needs it in the consensus module, as soon as a roster is received from the execution module?

The thing I like about it is the simplicity -- every module is already getting each round as it comes out of the Hashgraph, and it includes all the metadata and events including the roster information. So why have a separate path for sending the roster? We should only do that if it turns out that there is a correctness reason to do so.

Reply:
> Can you provide a good example of where this is meaningful?

It is highly advantageous for gossip to be able to accept and validate events as far into the future as possible. If gossip is limiting which events it can accept arbitrarily, we are sacrificing gossip efficiency.

Reply:
I'm not sure this is true. I am going to drop all events on the floor with a birth round too far in the future.

The argument I would make for having the "latest and greatest" is that if the neighbor assignments change rapidly, then having the most recent assignments may be necessary for me to progress as a node. For example, if Alice is on round 100 and everyone else is on 200, and there was a dramatic roster change somewhere between, then Alice will be trying to gossip with neighbors that refuse to talk with her, and won't be able to move forward, and will have to reconnect.

Reply:
> I am going to drop all events on the floor with a birth round too far in the future

This is the problem. If a node is declining to accept events because they are "too far in the future", but the node actually *does* have the required roster the accept and verify these events, then events aren't traveling as quickly through the network as they could be.

Being able to accept and verify events as far into the future as possible also helps smooth out bursty conditions.

Reply:
We want to drop events too far in the future, not only because we need the roster for it, but because this limits the number of events in the system if the node is falling behind. It cleanly solves the problem of giving backpressure for nodes that are under stress. So dropping events that are too far in the future is a feature, not a bug. Sometimes, optimizing one path to be as fast as possible leads to overall decreased performance and stability for the system as a whole. I think this is one of those times. Yes, a node that gets behind by enough rounds will stop using broadcast trees and start using sync, but if it is able to actually handle the events quickly enough to catch up, it will catch up pretty quickly, and then be back to broadcast.

>The argument I would make for having the "latest and greatest" is that if the neighbor assignments change rapidly, then having the most recent assignments may be necessary for me to progress as a node. For example, if Alice is on round 100 and everyone else is on 200, and there was a dramatic roster change somewhere between, then Alice will be trying to gossip with neighbors that refuse to talk with her, and won't be able to move forward, and will have to reconnect.

Actually, I take this back. This isn't a valid argument because the node cannot know of the roster until Execution has figured it out, and once it has it figured it out, the time it takes to apply it to the next consensus round and get it applied is generally very small. So I think the design as documented is both clean and desirable in terms of performance and stability.

Reply:
> We want to drop events too far in the future

Of course, we want to drop events that are *too far* in the future. The question is, what does "too far" mean? By routing the roster through the hashgraph module, we are greatly limiting our options.

*Maybe* it's the case that we don't want to handle any events with a birth round newer than the consensus round we are currently working on. If that's true, then routing the roster through the hashgraph is fine. But *maybe* it's the case that it's beneficial to gossip a couple rounds further into the future? Maybe the system is bursty enough that just having a bit more wiggle room makes things run smoothly. I don't think we can really know at this point. If we route rosters through hashgraph, then we don't have the option to even experiment with gossiping further into the future. 

On the other hand, if we pass the roster directly to gossip once receiving it from execution, we have the option to limit how far into the future we gossip as much as we want.

Reply:
I can say with high confidence that if we only give gossip the same address book that consensus is currently using, performance is going to get jittery and will likely see severe degradation in C2C. This is less of an issue when one node falls far behind its peers, but is critically important when nodes are only slightly out of sync with each other (this is the norm).

Suppose node A's consensus is 200 milliseconds behind node B's consensus. When B transitions from round `X` to round `X+1`, there will be a 200ms gap during which none of node B's new events can be sent to node A. This can be mitigated by allowing node A to gossip verifiable future events (aka giving gossip all address books computed by execution).

# Latent Gossip Issue

A word of warning about enabling gossip with birth rounds (not directly related to prior paragraphs). There is currently a major inefficiency in gossip when one peer is very far behind the other. This is because the node that is ahead iterates the entire hashgraph in order to find the relevent events to send.

This gets MUCH, MUCH worse when syncs are forced to be small. Using birth rounds to limit syncs, even if gossip can get verifiable future events, is going to force there to be small syncs. The net outcome of this is going to be that nodes that fall a little behind will never be able to catch up via gossip. I collected strong experimental data showing this phenomena, it's not just theory craft.

This issue is fixable, but the change is non-trivial and is going to require some deep changes to sync and sync data structures. Let me know if you'd like to discuss further, and I can describe the plan of attack I had originally planned on using to resolve this problem.

Reply:
>By routing the roster through the hashgraph module, we are greatly limiting our options.

Oh, why is this?

What I had in mind was that we'd pick a number `N`, say, 10. Let's say that we're about to work on round 103, but we're waiting for execution to give us a roster. It finally gives us a roster. We use it starting in round 103 + N. So I will have been open to receive events in birth rounds from 103, 104, 105, ... 112. I won't have received anything more future than that.

Once I produce round 103 (which should happen very quickly once we get the go-ahead from Execution), then I send to everyone else "Hey, you can now start accepting birth rounds up until 113, and if you do get one for 113, then use this roster".

The number N tells us how many future rounds we can accept events for. If the number is too large, we can get too many events in memory as the node falls behind. If it is too slow, then gossip gets janky. Also, if it is too large, it introduces latency between when the roster is available and when it can be applied. If someone is branching, it means they can branch for that many rounds.

Whether we send the roster to Hashgraph and then have it sent around, or whether we send it directly, has very little difference -- it is the time of computing a single round in the hashgraph. (It is very little time if all events are available and ready to go, and if all events are *not* available and ready to go, then there is no harm in delaying the distribution of the roster a bit, since I don't have a lot of future events coming in. If I did, I'd have had plenty of events to compute the round).

So I don't understand the concern. The design I've outlined allows for the Consensus system to have N number of future events, so it should be able to smooth out quite nicely. There is a tradeoff in "roster adoption latency", but that would be true no matter where we sent the roster. I believe the only extra delay is the time it takes to come to consensus for a single round, which is either very short because you have a ton of events handy, or it doesn't matter anyway because you have little/no future events to validate.

Reply:
> I can say with high confidence that if we only give gossip the same address book that consensus is currently using

This is not the proposal. The idea is that the round metadata has both the current roster and the future roster. But I think I didn't write that down, and it is only in my head. In which case, I emphatically agree. I think I didn't document this (probably at all). And that is probably the source of confusion! Let me take another whack at it.

Reply:
Is your plan to allow gossip to accept events for round `X` even if gossip/intake's latest address book is from round `X-10`?

This strategy has some major pitfalls. If we receive an event from a round where we don't have the roster, we can't validate that event. If we have an event that is not yet validated, then

- we may not send that event to anybody else (else risk an attacker being able to leverage honest nodes in spam attacks)
- we may not tell other nodes that we already have the event
    - if we tell other nodes not to send us this event and it turns out this event is invalid (e.g. the signature has been tampered with), then we will eventually discard the event and no honest nodes will resend the event. To get around this, we'd need a special `oops, I actually don't have that event I told you I had` protocol, which is a big complication.
    - as a result of being unable to tell others we have the event, any events received in this window will have very high duplicate rates
- we must hold onto and buffer data that we have not yet verified. Although not lethal in and of itself, this is a generally undesirable property to have. It should be a general goal to validate and discard invalid data ASAP.
    - spammer could send events from 10 rounds in the future for a node that doesn't exist. We can't throw them away because it's plausible that the node could be added to the address book by that time
    - spammer could send events with bad signatures that we have to hold onto for 10 rounds
    - etc.

When I describe future events, I split them into two categories:

- verifiable future events: events we have the address book for
- unverifiable future events: events we don't have the address book for

Consensus, event creation, etc. needs to be protected from all forms of future events. Gossip and event intake is safe to use verifiable future events. Any use of unverifiable future events in any layer of the system opens a huge attack surface against the protocol.

Reply:
>Is your plan to allow gossip to accept events for round X even if gossip/intake's latest address book is from round X-10?

This is not what I took away from @rbair23's explanation. Here is how I understand it:

The hashgraph module is trying to reach consensus on round 103, but is unable to because it's waiting for a roster from execution. The roster is it waiting on will become effective for round 113 (assuming N=10). This means gossip is currently accepting events up through round 112.

When the roster is received, it calculates consensus for round 103 (using the effective roster for that round) and sends the roster out to the rest of the interested modules. Gossip receives it and begins accepting events with birth round 113.

Reply:
> Is your plan to allow gossip to accept events for round X even if gossip/intake's latest address book is from round X-10?

No, you cannot accept events for rounds for which you do not have a roster.

Reply:
> The hashgraph module is trying to reach consensus on round 103, but is unable to because it's waiting for a roster from execution. The roster is it waiting on will become effective for round 113 (assuming N=10). This means gossip is currently accepting events whose birth round is 112 or less.

If I understand what you are saying, this changes the definition of birth round and what it means for the roster to "become effective". In such a system, you wouldn't be creating events with a birth round equal to what consensus is working on, you'd create events with a birth round equal to consensus's current round minus 10.

In this schema, there are actually two delays: the number of rounds between when execution finishes a roster and when consensus adopts the roster, and the delay between when consensus adopts the roster and when gossip/intake adopts the roster.

From my perspective, it's not obvious how this scheme is better than the original plan. It has all of the same downsides, but with additional complexity to boot.

Reply:
@poulok What I hear you describing is that consensus can't compute on round X until it knows what the roster will be for round X+N.  Execution, on every round, would have to communicate to consensus that it is done computing and whether or not a new roster is produced for N rounds into the future.   This doesn't sound like it gives any leeway for consensus to compute ahead of round handling.

Reply:
>This doesn't sound like it gives any leeway for consensus to compute ahead of round handling.

I think this is stated explicitly in this proposal. Consensus will not compute another round until Execution tells it to:
`In this design, we propose that the Hashgraph module never runs the consensus algorithm until it is told to do so from the Execution module`.

@rbair23, can you please expand on this previous comment and explain how the gossip module will accept events with birthround up to 112 when the roster for 103 is provided by Execution?
>What I had in mind was that we'd pick a number N, say, 10. Let's say that we're about to work on round 103, but we're waiting for execution to give us a roster. It finally gives us a roster. We use it starting in round 103 + N. So I will have been open to receive events in birth rounds from 103, 104, 105, ... 112. I won't have received anything more future than that.

Reply:
Would it be fair to describe it this way: 
* At the end of every handled round, execution calls back to consensus and indicates the roster to use N rounds in the future. 
* Consensus can compute on every round it has a known roster for. 
* If the last roster communication was that execution handled round X and gave a roster for round X+N, then consensus stops handling rounds once it handles X+N and waits for future rosters to be communicated.  

If consensus can't advance N rounds ahead of execution, then how many rounds can it safely advance?

Reply:
I'm updating with a better description. I had a conversation with Austin and Jasper. I think we're on the same page, I just really didn't explain it well.

![roster-changes](https://github.com/user-attachments/assets/41708f71-316c-4aa2-a284-a1ed32fa009e)

Here is what I'm planning to push later today:

```
#### Roster and Configuration Changes

When the roster (or any other network-wide configuration impacting Consensus) changes, Consensus must be updated in a
deterministic manner. This section discusses specifically how rosters are updated, but configuration in general would
use a similar path.

![Roster Changes](roster-changes.png)

Execution may hold many rounds. It will have a round that is currently being signed, one that is being hashed (which
happens before signing), one that is being handled, and any number of rounds it is holding awaiting execution. It is
very likely this "awaiting execution" buffer will be a very small number, perhaps as small as 1. This number of rounds
held in buffer is completely under the control of Execution. Consensus does not concern itself with rounds once they
are handed to Execution.

Consensus **only produces rounds on demand**. Execution will ask Consensus to produce the next round. Only then does
Consensus begin inserting events into the hashgraph and executing the hashgraph consensus algorithm. When any one event
is added to the hashgraph, it may produce 0 or more consensus rounds. Events are added until at least one consensus
round is produced.

The Execution module is responsible for defining the roster. When it asks Consensus for a new round, it also supplies
the new candidate roster. The Hashgraph module will decide in what round the candidate roster becomes active. Let us
define `N` as the number of rounds from the current consensus round at which the candidate roster will become active.
This number will be configured with the same value for all nodes in the network.

For example, suppose Execution has just finished handling round 10, and has buffered up round 11. The most recent round
that the Hashgraph module produced was round 11. Suppose `N` is 10. If the call to `nextRound` from Execution at the
end of round 10 were to supply a new Candidate Roster (CR). The Hashgraph module will assign the CR to become active as
of round 22 (`current_round + N + 1`). It will then work on producing consensus round 12. Once produced, metadata
indicating that the CR will be used for rounds starting at 22 is passed to Gossip, Event Intake, Event Creator, and
Execution. The current roster used for round 22 may also be included in this metadata.

There is a critical relationship between the value of N, the latency for adopting a new CR, and the number of future
events held in memory on the node. Remember that each event has a birth-round. Let us define another variable called
`Z`, such that `Z <= N`. Any event with `birth_round > current_round + z` is considered a "far future" event, and will
be dropped by the birth-round filtering logic in Event Intake. Any event with `birth_round > current_round` is just a
"future" event.

The smaller the value of `Z`, the fewer events are held in memory on the node. However, events with birth-rounds just
larger than `Z` will be dropped and have to be retrieved on demand with sync gossip, rather than through the broadcast
tree. Sync is slower and more laborious than broadcast gossip, so the efficiency of the node will go down with low
values of `Z`. Larger values of `Z` means more events in memory, but it also means more "smoothing" in gossip and
in handling any performance spikes.

On the other hand, the larger the value of `N`, the larger the latency between the round we know about a CR, and the
round at which it can be actually used. Ideally, the value of `N` would be small, like 3, but we may find that 3 is
too small a number of `Z`.

Each node may select its own value of `Z`, so long as it is less than or equal to `N`. But all nodes must use the same
value for `N`, or they will ISS since they will assign different rosters to different rounds. The value of `N` is not
defined here, but will need some investigation. Unfortunately, this number appears to require "just being chosen" and
cannot be deterministically dynamically computed based on network conditions.

Note: when the roster is applied to some future round `current_round + N + 1`, if such a round already exists in the
hashgraph, then that round and any future rounds must be removed and the events reapplied in order given the new
roster.

The Hashgraph module also includes a `state` section in the metadata of the round. This is used by the Execution layer
to persist the Consensus state, for reconnect and for restart. In the state are the judges for the round, and possibly
other information. When `initialize` is called, this same state is made available again to the Consensus node.
```

Reply:
>Note: when the roster is applied to some future round `current_round + N + 1`, if such a round already exists in the hashgraph, then that round and any future rounds must be removed and the events reapplied in order given the new roster.

I do not understand this statement. How can round `current_round + N + 1` already exist if `current_round` is the latest round consensus has calculated? Maybe I should ask if this definitely of `current_round` is correct.

Reply:
I may have off by one here. Edward suggested some better names for the round. Sometimes in the document I think I use current round as "most recent consensus round" and sometimes it means "the round we're currently trying to come to consensus on".

Given that, does the statement make sense? Or are you not sure about the "future rounds must be removed and the events reapplied in order given the new roster" part?

Reply:
>Or are you not sure about the "future rounds must be removed and the events reapplied in order given the new roster" part?

This is the part I don't follow.

Rounds only exist once consensus has calculated them. Consensus should never calculate a round for which it is not completely certain which roster should be used to calculate it. What do you mean by "future rounds must be removed and the events reapplied"?

Reply:
> I do not understand this statement. How can round current_round + N + 1 already exist if current_round is the latest round consensus has calculated? Maybe I should ask if this definition of current_round is correct.

So let's assume that `N` is some number, like 4. So suppose I have:

```
 | Round 102 |   <-- current vote collection round, the round that collects votes from the judges in 101
 | Round 101 |    <-- current voting round, the round with judges collecting votes on round 100 judges
 | Round 100 |   <-- pending consensus round, the round I'm trying to get consensus on
 | Round 99  |    <-- latest consensus round, has now come to consensus
```

In this case, if we were given a roster and `N` was 3 (and if we count starting from 99), then we would say "this roster R1 will be used as of round 103.

In the vast majority of cases, we will only ever see 3 rounds in hashgraph that we haven't achieved consensus on yet. But it is possible that round 102 isn't able to get enough nodes to see enough events to come to consensus on 100. In that case, we will get round 103 created. And this in theory could continue for a while, which is why we have coin rounds. And the hashgraph paper shows that the probability of multiple coin rounds is incredibly low, and in fact will eventually conclude with probability 1.

Now, what happens if we actually already had a round 104, or 105, or 106 at the time we received the new roster `R1`? In that case, I would try to assign `R1` to some round that already exists in the hashgraph. So I would need to throw that round and all subsequent rounds away, and then reapply those events in topological order, given the new roster `R1`.

If `N` was something more like `10`, then it is extremely unlikely that this code path would ever be executed. But if it were, we'd need to throw those rounds away and reapply the events in topological order with the correct roster.

We cannot just put the roster at some dynamic point in the future based on what rounds I have already created in my hashgraph, because this can be nondeterministic across nodes in the network.

Reply:
>Rounds only exist once consensus has calculated them. Consensus should never calculate a round for which it is not completely certain which roster should be used to calculate it. What do you mean by "future rounds must be removed and the events reapplied"?

Ah, no. Rounds are created before consensus. See the "divideRounds" pseudocode method in the whitepaper for details. In fact, you need 3 rounds to exist for the first to come to consensus, but in some cases, you can end up with many more rounds than that, none of which have come to consensus.

Reply:
OK, this goes back to there being two kinds of rounds: consensus rounds and voting rounds (used to be round created). Consensus round never changes once calculated. Voting round can change, which I believe, happens when the metadata of non-consensus events is recalculated. I do not think the events are _removed and re-added_ but this could just be a terminology detail.

At some point there was an effort to get rid of this idea of there being two kinds of rounds and I always found it to be very confusing. Consensus round is immutable whereas round created is mutable, and the two values do not have to match. But I suppose with this single definition of round, it is mutable for a time and then becomes immutable.

Reply:
As best as I understand it, the only relationship between the terms "round" (as we normally use it) and "voting round" is that they have unfortunately similar names. Perhaps "voting round" should be renamed... some suggestions below.

- election
- election cycle
- voting cycle
- voting iteration
- entmoot

Reply:
Talking with Leemon, the terms as I understand them are:

- Birth Round
- Voting Round
- Consensus Round

Reply:
Much cleaner.